plugins/pulp_docker/plugins/distributors/publish_steps.py

```diff
@@ -220,3 +222,126 @@ def process_main(self):
         misc.mkdir(os.path.dirname(self.app_publish_location))
         with open(self.app_publish_location, 'w') as app_file:
             app_file.write(json.dumps(redirect_data))
+
+
+class PublishTagsForRsyncStep(RSyncFastForwardUnitPublishStep):
+
+    def __init__(self, step_type, model_classes, repo_content_unit_q=None, repo=None,
+                 config=None, remote_repo_path=None, published_unit_path=None, unit_fields=None):
+
+        super(PublishTagsForRsyncStep, self).__init__(step_type, model_classes,
+                                                      repo_content_unit_q=repo_content_unit_q,
+                                                      repo=repo, config=config,
+                                                      remote_repo_path=remote_repo_path,
+                                                      published_unit_path=published_unit_path,
+                                                      unit_fields=unit_fields)
+        self._tag_names = set()
+
+    def process_main(self, item=None):
+        """
+        Create the manifest tag relative links.
+
+        :param item: The tag to process
+        :type  item: pulp_docker.plugins.models.Tag
+        """
+        manifest = models.Manifest.objects.get(digest=item.manifest_digest)
+        filename = item.name
+        symlink = self.make_link_unit(manifest, filename, self.get_working_dir(),
+                                      self.remote_repo_path,
+                                      self.get_config().get("remote")["root"],
+                                      self.published_unit_path)
+        self.parent.symlink_list.append(symlink)
+        self._tag_names.add(item.name)
+
+    def finalize(self):
+        """
+        Write the Tag list file so that clients can retrieve the list of available Tags.
+        """
+        tags_path = os.path.join(self.parent.get_working_dir(), '.relative', 'tags')
+        misc.mkdir(tags_path)
+        with open(os.path.join(tags_path, 'list'), 'w') as list_file:
+            tag_data = {
+                'name': configuration.get_repo_registry_id(self.get_repo(), self.get_config()),
+                'tags': list(self._tag_names)}
+            list_file.write(json.dumps(tag_data))
+        # We don't need the tag names anymore
+        del self._tag_names
+
+
+class DockerRsyncPublisher(Publisher):
+
+    REPO_CONTENT_TYPES = [constants.IMAGE_TYPE_ID, constants.BLOB_TYPE_ID,
+                          constants.MANIFEST_TYPE_ID]
+
+    REPO_CONTENT_MODELS = [models.Blob, models.Manifest, models.Image]
+
+    def _get_predistributor(self):
+        """
+        Returns distributor which repo has to be published in before
```

Comment:
I find this description a bit difficult to understand. Perhaps s/in/with/ would help, since I think of publishing _with_ a distributor, not _in_ one. I'm also not sure what the words "content generator" at the end are telling me.